doc : Document new attributs, AsGrid and AsFilter (#267)
Sylius/Stack: docs/grid/custom_filter.md

To add a new filter, we need to create an appropriate class and form type.

## Using attributes

> {% code title="src/Grid/Filter/SuppliersStatisticsFilter.php" lineNumbers="true" %}

Reviewer: I think it should be in singular form, same as at most parts of the sylius.

```suggestion
{% code title="src/Grid/Filter/SupplierStatisticsFilter.php" lineNumbers="true" %}
```

Author: This is a copy of the previous documentation, if we change this we should change the "legacy way" documentation too.

```php

> <?php
> 
> namespace App\Grid\Filter;

Reviewer: Since it's a full php code and not just snippet -- I think we should use strict over here.

```suggestion
<?php

declare(strict_types=1);

namespace App\Grid\Filter;
```

use App\Form\Type\Filter\SuppliersStatisticsFilterType;
use Sylius\Bundle\GridBundle\Doctrine\DataSourceInterface;
use Sylius\Component\Grid\Attribute\AsFilter;
use Sylius\Component\Grid\Filtering\FilterInterface;

#[AsFilter(
    formType: SuppliersStatisticsFilterType::class,
    type: 'suppliers_statistics',

>     template: 'templates/Grid/Filter/suppliers_statistics.html.twig'

Reviewer: Sylius 2.0 and stack uses new file structure for templates dir with lowercase names.

```suggestion
    template: 'templates/grid/filter/suppliers_statistics.html.twig'
```

Is `supplier statistics` related to dashboard or some other page? I thinking of what better path will suite it, rather than root `/grid/filter/`.

)]
class SuppliersStatisticsFilter implements FilterInterface
{
    public function apply(DataSourceInterface $dataSource, $name, $data, array $options = []): void
    {
        // Your filtering logic.
        // $data['stats'] contains the submitted value!
        $queryBuilder = $dataSource->getQueryBuilder();
        $queryBuilder
            ->andWhere('stats = :stats')
            ->setParameter(':stats', $data['stats'])
        ;
    
        // You can leverage the ExpressionBuilder to apply driver-agnostic filters to the data source.
        // Combined with restrict(), it provides query builder–style functionality for grid filters.
        $dataSource->restrict($dataSource->getExpressionBuilder()->equals('stats', $data['stats']));
    }
}
```
{% endcode %}

And the form type:

> {% code title="src/Grid/Filter/SuppliersStatisticsFilterType.php" lineNumbers="true" %}

Reviewer: ```suggestion
{% code title="src/Form/Type/Filter/SuppliersStatisticsFilterType.php" lineNumbers="true" %}
```

```php

> <?php
> 
> namespace App\Form\Type\Filter;

Reviewer: ```suggestion
<?php

declare(strict_types=1);

namespace App\Form\Type\Filter;
```

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class SuppliersStatisticsFilterType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options): void
    {
        $builder->add(

>             'stats',

Reviewer: I understand `stats` existed in the examples before, but it's hard to read. Does it mean days, count, limit? A thought for future to improve it somehow.

            ChoiceType::class,

>             ['choices' => range($options['range'][0], $options['range'][1])]

Reviewer: This line doesn't look good as official example, it should be at least like so.

```suggestion
            ['choices' => range($options['from'], $options['to'])]
```

        );
    }

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver
            ->setDefaults([
                'range' => [0, 10],
            ])
            ->setAllowedTypes('range', ['array'])
        ;
    }
}
```
{% endcode %}

Create a template for the filter, similar to the existing ones:

{% code title="templates/Grid/Filter/suppliers_statistics.html.twig" lineNumbers="true" %}
```twig

> {% form_theme form '@SyliusUi/Form/theme.html.twig' %}

Reviewer: That's a sylius v1 code, stack has it's own bootstrap form theme.

This one I think 
https://github.com/Sylius/Stack/blob/main/src/BootstrapAdminUi/templates/shared/form_theme.html.twig

Reviewer 2: yes, this code is outdated. @diimpp is right.

## Using attributes

{% code title="src/Grid/Filter/SuppliersStatisticsFilter.php" lineNumbers="true" %}
```php
<?php

namespace App\Grid\Filter;

use App\Form\Type\Filter\SuppliersStatisticsFilterType;
use Sylius\Bundle\GridBundle\Doctrine\DataSourceInterface;
use Sylius\Component\Grid\Attribute\AsFilter;
use Sylius\Component\Grid\Filtering\FilterInterface;

#[AsFilter(
    formType: SuppliersStatisticsFilterType::class,
    type: 'suppliers_statistics',
    template: 'templates/Grid/Filter/suppliers_statistics.html.twig'
)]
class SuppliersStatisticsFilter implements FilterInterface
{
    public function apply(DataSourceInterface $dataSource, $name, $data, array $options = []): void
    {
        // Your filtering logic.
        // $data['stats'] contains the submitted value!
        $queryBuilder = $dataSource->getQueryBuilder();
        $queryBuilder
            ->andWhere('stats = :stats')
            ->setParameter(':stats', $data['stats'])
        ;
    
        // You can leverage the ExpressionBuilder to apply driver-agnostic filters to the data source.
       // Combined with restrict(), it provides query builder–style functionality for grid filters.

>         // $data['stats'] contains the submitted value!

Reviewer 3: ```suggestion
``` 
(already mentioned just above)

And the form type:

{% code title="src/Grid/Filter/SuppliersStatisticsFilterType.php" lineNumbers="true" %}
```php
<?php

namespace App\Form\Type\Filter;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class SuppliersStatisticsFilterType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options): void
    {
        $builder->add(
            'stats',
            ChoiceType::class,
            ['choices' => range($options['range'][0], $options['range'][1])]
        );
    }

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver
            ->setDefaults([
                'range' => [0, 10],
            ])
            ->setAllowedTypes('range', ['array'])
        ;
    }
}
```
{% endcode %}

Create a template for the filter, similar to the existing ones:

{% code title="templates/Grid/Filter/suppliers_statistics.html.twig" lineNumbers="true" %}
```twig
{% form_theme form '@SyliusUi/Form/theme.html.twig' %}

{{ form_row(form) }}
```
{% endcode %}

> Now you can use your new filter type in the grid configuration!

Reviewer 3: ```suggestion
Your custom filter is now ready to be plugged into the grid configuration using its service alias or FQCN.
```
Here, I think we should show both examples so I would also add something like :
```php
 ->addFilter(
                Filter::create('stats', 'suppliers_statistics')
                    ->setFormOptions(['range' => [0, 100]])
            )
```
 either as a new tab or below inside the same grid with a comment

    {
        $builder->add(
            'stats',
            ChoiceType::class,
            ['choices' => range($options['from'], $options['to'])]
        );
    }

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver
            ->setDefaults([
                'range' => [0, 10],
            ])
            ->setAllowedTypes('range', ['array'])
        ;
    }
}
```
{% endcode %}

Create a template for the filter, similar to the existing ones:

{% code title="templates/grid/filter/suppliers_statistics.html.twig" lineNumbers="true" %}
```twig
{% form_theme form '@SyliusUi/Form/theme.html.twig' %}

{{ form_row(form) }}
```
{% endcode %}

Your custom filter is now ready to be plugged into the grid configuration using its service alias or FQCN.

{% code title="src/Grid/TournamentGrid.php" lineNumbers="true" %}
```php
<?php

declare(strict_types=1);

namespace App\Grid;

use App\Entity\Tournament;
use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;

> #[AsGrid('app_tournament', Tournament::class)]

Reviewer 2: please use named arguments.
